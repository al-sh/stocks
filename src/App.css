@import './components/analytics/Analytics.css';
@import './components/menu/Menu.css';
@import './components/common/throbber/Throbber.css';
@import './components/RTable/RTable.css';

.App {
  margin: auto;
  width: 1000px;    
  text-align: left;
}
