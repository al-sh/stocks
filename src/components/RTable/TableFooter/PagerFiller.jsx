import React from 'react';

class PagerFiller extends React.Component {
  render() {
    return (
      <span className="pagerFiller">
        ...
      </span>
    );
  }
}

export default PagerFiller