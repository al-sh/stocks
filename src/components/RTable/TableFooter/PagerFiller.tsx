import React from 'react';

const PagerFiller: React.FC = () => {
  return (
    <span className="pagerFiller">
      ...
    </span>
  );
}

export default PagerFiller