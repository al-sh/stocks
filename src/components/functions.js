/*function formatDate(dt){ 
  const lPadZeros = value => { return value.toString().length===1?'0'+value:value }
  
  const year = dt.getFullYear();
  const month = lPadZeros( dt.getMonth()+1 );
  const day = lPadZeros( dt.getDate() );

  const hours = lPadZeros( dt.getHours() );
  const minutes = lPadZeros( dt.getMinutes() );
  const seconds = lPadZeros( dt.getSeconds() );

  return year + '-' + month + '-' + day + ' ' + hours + ':'+ minutes + ':' + seconds
}

function sendPOST(url, data) {
  return new Promise(function(resolve, reject) {
    var request = new XMLHttpRequest();
    request.open("POST", url, true);
    request.addEventListener("load", function() {
      if (request.status < 400)
        resolve(request.response);
      else
        reject(new Error("Request failed: " + request.statusText));
    });
    request.addEventListener("error", function() {
      reject(new Error("Network error"));
    });
    console.log('POST Data:', data)
    request.send(JSON.stringify(data));
  });
}*/

function findEl(arr, id){
  for(let i=0; i<arr.length; i++){
    if(arr[i].id===id){
      return i;  
    }
  }
}

function deleteItemFromArray(arr, itemId){
  const indexToDel = findEl(arr, itemId); 
  arr.splice(indexToDel, 1);
}

function getNewId(scData){
  let newId = 0;

  for(let i=0; i<scData.length; i++){
    if ( parseInt(scData[i].id)> newId){
      newId = scData[i].id;
    }
  }
  return newId+1;
}

function getPageRows(items, pageNumber, rowsinPage){  
  const rowsCount = items.length;
  
  const firstRowNum = pageNumber * rowsinPage; 
  let lastRowNum = (pageNumber + 1) * rowsinPage; 
  if (lastRowNum>rowsCount) lastRowNum = rowsCount;
  
  let rows = [];
  for(let i=firstRowNum; i < lastRowNum; i++){
    rows.push(items[i]);
  }

  return rows;
}

//на входе - json, в котором указано в каком порядке сортируется и сам scrollerData
//на выходе - отсортированный scrollerData (изменяет текущий)
//const mSort = (scData, sortParams) => { 
function mSort(scData, sortParams) {   
  const defaultCompare = (a, b, fieldName, ascOrder) => {    
    if (a[fieldName] > b[fieldName]) { 
      return ascOrder?1:-1;
    }
    if (a[fieldName] < b[fieldName]) {
      return ascOrder?-1:1;         
    }  
    return 0; 
  }

  const floatCompare = (a, b, fieldName, ascOrder) => {  
    const first = isFinite(parseFloat(a[fieldName]))?parseFloat(a[fieldName]):0;
    const second = isFinite(parseFloat(b[fieldName]))?parseFloat(b[fieldName]):0;

    if (first > second) { 
      return ascOrder?1:-1;
    }
    if (first < second) {
      return ascOrder?-1:1;         
    }  
    return 0; 
  }

  const singleCompare = (a, b, currentSortParam) => {
    if (currentSortParam.type === 'float'){
      return floatCompare(a, b, currentSortParam.field, currentSortParam.ascOrder);
    }
    if (currentSortParam.type === 'default'){
      return defaultCompare(a, b, currentSortParam.field, currentSortParam.ascOrder);
    }
  }  
  
  const compareFunction = (a, b) => {
    let currentSortParam;

    for (let i=0; i<sortParams.length; i++){
      currentSortParam = sortParams[i];
      const currentResult = singleCompare(a, b, currentSortParam);
      if (currentResult!==0) {
        return currentResult;
      } 
    }
    return 0;
  } 

  return scData.sort(compareFunction);
}

function maskAmount(x) {
  if (!isFinite(x)) {
    return '';
  }

  var parts = x.toString().split(".");
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, " ");
  parts[1] = parts[1]?parts[1].substring(0, 2):'00';
  return parts.join(".");
}

/*async function getDataFromBack(url) {
  return new Promise(function(resolve, reject) {
    var request = new XMLHttpRequest();
    request.open("GET", url, true);
    request.addEventListener("load", function() {
      if (request.status < 400)
        resolve(request.response);
      else
        reject(new Error("Request failed: " + request.statusText));
    });
    request.addEventListener("error", function() {
      reject(new Error("Network error"));
    });
    request.send();
  });
}*/

async function getDataFromBack(url) {
  let response = await fetch(url);
  if (response.ok) {
    return await response.json();
  } else {
    console.log("Ошибка HTTP: " + response.status);
    return false;
  }
}

export { findEl, mSort, getNewId, getPageRows, deleteItemFromArray, maskAmount, getDataFromBack };