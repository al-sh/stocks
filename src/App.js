import React from 'react';
import { connect } from 'react-redux';
import './App.css';
import MainMenu from './components/menu/MainMenu';
import RTable from './components/RTable/RTable';
import Throbber from './components/common/throbber/Throbber';

class App extends React.Component { 
  render() {  
    return (         
      <div className="App">      
        <MainMenu />
        { this.props.showThrobber ? <Throbber /> : <RTable /> }           
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {  
    showThrobber: state.showThrobber
  }
}

export default connect(
  mapStateToProps,
  null //mapDispatchToProps
)(App)
