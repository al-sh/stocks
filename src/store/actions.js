import { testPortfolios, testTransfers } from './testData'
import { getDataFromBack } from '../components/functions';

const testDataWarnText = 'Отсутствует подключение к БД, показаны тестовые данные';

export function toggleVisualSort(fieldName, sortType, ctrlPressed) {
    return { type: 'SORT.TOGGLE', field: fieldName, sortType: sortType, ctrlPressed: ctrlPressed }
}

function toggleThrobber(dispatch, visible) {
    dispatch({
        type: 'TOGGLE_THROBBER',
        showThrobber: visible
    });
}

const sections = {
    portfolios: {
        backUrl: '/getportfolios',
        testItems: testPortfolios,
        actionType: 'SWITCH.PORTFOLIOS'
    },
    transfers: {
        backUrl: '/transfers.getall',
        testItems: testTransfers,
        actionType: 'SWITCH.TRANSFERS'
    }
}

export async function actSwitchToSection(sectionName, dispatch) {
    toggleThrobber(dispatch, true);

    const selectedSection = sections[sectionName];
    let newItems = await getDataFromBack(selectedSection.backUrl);
    if (!newItems) {
        console.warn(testDataWarnText);
        newItems = selectedSection.testItems;
    }
    console.log(newItems);

    dispatch({
        type: selectedSection.actionType,
        items: newItems
    });
    toggleThrobber(dispatch, false);
}