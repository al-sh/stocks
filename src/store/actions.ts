import { testPortfolios, testTransfers } from './testData'
import { getDataFromBack } from '../components/functions';

const testDataWarnText = 'Отсутствует подключение к БД, показаны тестовые данные';

export function toggleVisualSort(fieldName: string, sortType: string, ctrlPressed: boolean) {
    return { type: 'SORT.TOGGLE', field: fieldName, sortType: sortType, ctrlPressed: ctrlPressed }
}

function toggleThrobber(dispatch: any, visible: boolean) {
    dispatch({
        type: 'TOGGLE_THROBBER',
        showThrobber: visible
    });
}

const sections: any = {
    portfolios: {
        backUrl: '/getportfolios',
        testItems: testPortfolios,
        actionType: 'SWITCH.PORTFOLIOS'
    },
    transfers: {
        backUrl: '/transfers.getall',
        testItems: testTransfers,
        actionType: 'SWITCH.TRANSFERS'
    }
}

export async function actSwitchToSection(sectionName: string, dispatch: any) {
    toggleThrobber(dispatch, true);

    const selectedSection = sections[sectionName];
    let newItems = await getDataFromBack(selectedSection.backUrl);
    if (!newItems) {
        console.warn(testDataWarnText);
        newItems = selectedSection.testItems;
    }
    console.log(newItems);

    dispatch({
        type: selectedSection.actionType,
        items: newItems
    });
    toggleThrobber(dispatch, false);
}