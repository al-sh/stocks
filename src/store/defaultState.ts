import { testPortfolios } from './testData'
import { IState } from './interfaces'
import { maskAmount } from '../components/functions';

export const defaultState: IState = {
    analytics: {
        currentPortfolioValue: 400000
    }, 
    currentSection: 'portfolios',
    columns: [
        {
            width: '200px',
            field: 'name',
            text: 'Название',
            visible: true
        },
        {
            width: '160px',
            field: 'isActive',
            text: 'Статус',
            format: (isActive: boolean) => { return isActive ? 'Активный' : 'Не активный'; },
            visible: true
        },
        {
            width: '150px',
            field: 'amount',
            text: 'Сумма',
            type: 'float',
            headerClassName: 'header-amount',
            style: { textAlign: 'right', paddingRight: '15px' },
            format: maskAmount,
            visible: true
        }
    ],
    items: testPortfolios,
    selectedItems: [],
    settings: {
        rowsinPage: 5,
        activePage: 0,
        withCheckBoxes: false,
        showDisplaySettings: false,
        showAddEditBlock: false,
        editItem: null,
        shiftClickItemId: null,
        toolBarRowId: null
    },
    showThrobber: false,
    sortParams: [{ field: "name", type: "default", ascOrder: true }]
};