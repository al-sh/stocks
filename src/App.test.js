import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';

import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { defaultState } from './store/defaultState'
import rootReducer from './store/reducers/mainReducer'

import { mSort } from './components/functions';

import PagerFiller from './components/RTable/TableFooter/PagerFiller';
import renderer from 'react-test-renderer';

const rootDiv = document.createElement('div');
mSort(defaultState.items, defaultState.sortParams);
let preLoadedState = defaultState;

const store = createStore(
	rootReducer, preLoadedState
);

describe('main tests', () => {
	it('renders without crashing', () => {
		ReactDOM.render(
			<Provider store={store}>
				<App />
			</Provider>,
			rootDiv
		);
	});

	
	//test for simple component, without redux
	it("PagerFiller check", () => {
		const component = renderer.create(<PagerFiller />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		/* 
		// manually trigger the callback
			tree.props.onMouseEnter();
			// re-rendering
			tree = component.toJSON();
			expect(tree).toMatchSnapshot();

			// manually trigger the callback
			tree.props.onMouseLeave();
			// re-rendering
			tree = component.toJSON();
			expect(tree).toMatchSnapshot();
   */
	});

	//redux tests: https://habr.com/ru/post/340514/

	it('unmount without crashing', () => {
		ReactDOM.unmountComponentAtNode(rootDiv);
	});
});