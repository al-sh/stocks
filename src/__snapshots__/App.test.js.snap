// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main tests PagerFiller check 1`] = `
<span
  className="pagerFiller"
>
  ...
</span>
`;
